src/atomWithSearchParams.ts

```diff
@@ -0,0 +1,80 @@
+import type { SetStateAction, WritableAtom } from 'jotai/vanilla';
+import { atom } from 'jotai/vanilla';
+import { atomWithLocation } from './atomWithLocation';
+
+// Create an atom for managing location state, including search parameters.
+const locationAtom = atomWithLocation();
+
+/**
+ * Creates a writable Jotai atom to manage a search parameter in the URL.
+ *
+ * @template T - The type of the s parameter value (e.g., string or number).
+ * @param key - The name of the search parameter to manage.
+ * @param defaultValue - The default value for the search parameter if not present in the URL.
+ * @returns A writable atom for reading and updating the search parameter.
+ */
+export const atomWithSearchParams = <T>(
+  key: string,
+  defaultValue: T,
+): WritableAtom<T, [SetStateAction<T>], void> => {
+  /**
+   * Resolves the value of a search parameter based on its type.
+   *
+   * @param value - The raw string value from the URL (or `null` if absent).
+   * @returns The resolved value cast to type `T`.
+   */
+  const resolveDefaultValue = (value: string | null | undefined): T => {
+    // If the value is null or undefined, return the default value.
+    if (value === null || value === undefined) {
+      return defaultValue;
+    }
+
+    // If the default value is a number, attempt to parse the value as a number.
```

Comment:
On second look, this feels too heuristic.

Can we simply accept only string type?

Such as:
```ts
export const atomWithSearchParams = (
  key: string,
  defaultValue: string,
): WritableAtom<string, [SetStateAction<string>], void> => {
```

Reply:
If necessary, we should provide another util to convert string atom to number atom that can be used optionally.

Reply:
Although it seems heuristic to accept multiple types, allowing numbers or other types is convenient for state updates, like incrementing the page number dynamically `setPage(c => c + 1)`.

Forcing the value to be a string would lead to unexpected results,
such as `?page=11` instead of the intended numerical operation `const pageAtom = atomWithSearchParams("page", 1)`.

Reply:
Please consider some explicit ways something like:
```js
const pageAtom = atomWithSearchParams("key", 123, {
  type: 'number',
  // optionally custom parser?
});
```

Reply:
I have added JSON serialization functionality for scenarios where `defaultValue` might be of type `object`. Based on my use case, I don't think there's a need to add an additional custom parser. I’ve listed the following use cases for reference, but please let me know if I misunderstood your point.

URL: `?page=4&uuid=1&isAuth=true&obj=%7B%22count%22%3A3%2C%22data%22%3A%5B4%2C5%2C6%5D%7D`

```
const jsonData = {
  count: 1,
  data: [1, 2, 3],
};

const pageAtom = atomWithSearchParams("page", 1);
const uuidAtom = atomWithSearchParams("uuid", "");
const isAuthAtom = atomWithSearchParams("isAuth", false);
const objectAtom = atomWithSearchParams("obj", jsonData);

function App() {
  const [page, setPage] = useAtom(pageAtom);
  const [uuid, setUuid] = useAtom(uuidAtom);
  const [isAuth, setIsAuth] = useAtom(isAuthAtom);
  const [object, setObject] = useAtom(objectAtom);

  console.log("render", object);
  return (
    <>
      <p>{page}</p>
      <button onClick={() => setPage((c) => c + 1)}>Set Page</button>

      <p>{uuid}</p>
      <button onClick={() => setUuid((c) => c + 1)}>Set UUID</button>

      <p>{isAuth ? "true" : "false"}</p>
      <button onClick={() => setIsAuth((c) => !c)}>Set IS Auth</button>

      <p>{JSON.stringify(object)}</p>
      <button
        onClick={() =>
          setObject((c) => {
            return { ...c, count: c.count + 1, data: [4, 5, 6] };
          })
        }>
        Set Page
      </button>
    </>
  );
}
```

Reply:
I still think that simply accepting only the string type is not as flexible. Using additional utilities and parsers to convert the string into other types would be quite cumbersome.

Reply:
Sorry, I made you slightly misunderstand my requirement. My concern is lying types. I thought explicit "type" option would help, but your solution of checking the runtime type of defaultValue would work too if we were to support only primitive types.

For complex types (aka object), users can use "string" and parse/stringify on their own.

> I still think that simply accepting only the string type is not as flexible.

OK, let's only support some primitive types. What's your suggestion?

Reply:
I agree with your idea. For the current use case, primitive types are fully sufficient. If there are special requirements in the future, we can add an options parser when needed.